src/common/download.cpp

```diff
@@ -40,27 +41,86 @@
 
 namespace tools
 {
-  static bool download_thread(const std::string &path, const std::string &url)
+  struct download_thread_control
   {
+    const std::string path;
+    const std::string uri;
+    std::function<void(const std::string&, const std::string&, bool)> result_cb;
+    std::function<bool(const std::string&, const std::string&, size_t, ssize_t)> progress_cb;
+    bool stop;
+    bool stopped;
+    bool success;
+    std::shared_ptr<boost::thread> thread;
+    boost::mutex mutex;
+
+    download_thread_control(const std::string &path, const std::string &uri, std::function<void(const std::string&, const std::string&, bool)> result_cb, std::function<bool(const std::string&, const std::string&, size_t, ssize_t)> progress_cb):
+        path(path), uri(uri), result_cb(result_cb), progress_cb(progress_cb), stop(false), stopped(false), success(false) {}
+    void lock() { mutex.lock(); }
+    void unlock() { mutex.unlock(); }
+  };
+
+  static void download_thread(download_thread_control *control)
+  {
+    static std::atomic<unsigned int> thread_id(0);
+
+    MLOG_SET_THREAD_NAME("DL" + std::to_string(thread_id++));
+
+    epee::misc_utils::auto_scope_leave_caller scope_exit_handler = epee::misc_utils::create_scope_leave_handler([&control](){ delete control; });
     try
     {
-      MINFO("Downloading " << url << " to " << path);
+      control->lock();
```

Comment:
A number of these calls might be better with `boost::lock_guard<boost::mutex>`. There might be some ways to cause deadlock on the caller if an exception is thrown before the `unlock`, but I haven't dug too deeply.